### DISCUSSION TITLE
Location of platformio/framework-arduino-megaavr-megatinycore package git repo.

### DISCUSSION THREAD

**Discussion by ieb** [SpenceKonde/megaTinyCore#877] [Q&A]
Sorry to have to ask another PlatformIO question here.

Where is the source code for https://registry.platformio.org/tools/platformio/framework-arduino-megaavr-megatinycore which references this git repo. 

When that package is installed it uses files from https://github.com/SpenceKonde/megaTinyCore/tree/master/megaavr mixed in with files from the root and some files that dont exists, which makes me think the pointer to this repo is incorrect.  I want to ask the maintainer to create a 2.6.4 version of the package.

It is possible to install manually by checking out 2.6.4 locally, copying ~/.platformio/packages/ramework-arduino-megaavr-megatinycore (2.5.11) into ~/.platformio/packages/ramework-arduino-megaavr-megatinycore@2.6.4, replace the 2.5.11 files with the 2.6.4 files (I used symlinks) and adjusting the .poipm and package.json files.

Then to use.

        platform = atmelmegaavr
        platform_packages = 
             platformio/framework-arduino-megaavr-megatinycore@2.6.4
             
 Unfortunately referencing a git repo fails as the version number isn't in the expected format. (too many '.'s)

**Comment by SpenceKonde** [comment]
Please be more specific, can you list which files are treated in what ways? And is this a megaTinyCore bug or a platformio bug?

**Comment by ieb** [comment]
Here were the "errors" I encountered. It might be expected behaviour rather than any bug.

TLDR, should there be a package.json in megaTinyCore git repo to allow platform_packages to reference the git repo directly ?

Long version:

Starting from a default megaTinyCore install with Platform IO that is currently at megaTinyCore 2.5.11 I wanted to upgrade to 2.6.4 to get the improvements to Serial classes.

So I tried 

```
[env:attiny3224]
platform = atmelmegaavr
platform_packages = 
     platformio/framework-arduino-megaavr-megatinycore @ https://github.com/SpenceKonde/megaTinyCore.git
board = attiny3224
```
which results a report that the `package.json` is missing, hence this asking where the git repo that was used to build the 2.5.11 package is.

```
Processing attiny3224 (platform: atmelmegaavr; board: attiny3224; framework: arduino)
--------------------------------------------------------------------------------
Tool Manager: Installing git+https://github.com/SpenceKonde/megaTinyCore.git
git version 2.9.3 (Apple Git-75)
Cloning into '/Users/ieb/.platformio/.cache/tmp/pkg-installing-wfmf2lso'...
Error: Could not find one of 'package.json' manifest files in the package
[Finished in 6.3s with exit code 1]
[cmd: ['/usr/local/bin/pio', '-f', '-c', 'sublimetext', 'run', '--target', 'upload']]
```

Then I tried patching the 2.4.6 tag in a branch of my own 

```
[env:attiny3224]
platform = atmelmegaavr
platform_packages = 
     platformio/framework-arduino-megaavr-megatinycore @ https://github.com/ieb/megaTinyCore.git#2.6.4-pio
board = attiny3224
framework = arduino
```


resulting in exposing a bug in platforms/atmelmegaavr/builder/frameworks/arduino.py  which I can report there. (can't parse `2.6.4+sha.9025674` as a package version, too many '.'s)

```
Processing attiny3224 (platform: atmelmegaavr; board: attiny3224; framework: arduino)
--------------------------------------------------------------------------------
Tool Manager: Installing git+https://github.com/ieb/megaTinyCore.git#2.6.4-pio
git version 2.9.3 (Apple Git-75)
Cloning into '/Users/ieb/.platformio/.cache/tmp/pkg-installing-sl3txi8k'...
Tool Manager: framework-arduino-megaavr-megatinycore@2.6.4+sha.9025674 has been installed!
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/atmelmegaavr/attiny3224.html
PLATFORM: Atmel megaAVR (1.6.0) > ATtiny3224
HARDWARE: ATTINY3224 16MHz, 3KB RAM, 32KB Flash
PACKAGES: 
 - framework-arduino-megaavr-megatinycore @ 2.6.4+sha.9025674 
 - tool-avrdude-megaavr @ 3.60300.220118 (6.3.0) 
 - toolchain-atmelavr @ 3.70300.220127 (7.3.0)
2.6.4+sha.9025674
ValueError: too many values to unpack (expected 3):
  File "/Users/ieb/.platformio/penv/lib/python3.9/site-packages/platformio/builder/main.py", line 188:
    env.SConscript("$BUILD_SCRIPT")
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/Users/ieb/.platformio/platforms/atmelmegaavr/builder/main.py", line 148:
    target_elf = env.BuildProgram()
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/Users/ieb/.platformio/penv/lib/python3.9/site-packages/platformio/builder/tools/platformio.py", line 62:
    env.ProcessProgramDeps()
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/Users/ieb/.platformio/penv/lib/python3.9/site-packages/platformio/builder/tools/platformio.py", line 122:
    env.BuildFrameworks(env.get("PIOFRAMEWORK"))
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/Users/ieb/.platformio/penv/lib/python3.9/site-packages/platformio/builder/tools/platformio.py", line 328:
    SConscript(env.GetFrameworkScript(name), exports="env")
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 660:
    return method(*args, **kw)
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/Users/ieb/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/Users/ieb/.platformio/platforms/atmelmegaavr/builder/frameworks/arduino.py", line 85:
    major, minor, patch = package_version.split(".")
========================== [FAILED] Took 9.83 seconds ==========================
[Finished in 11.5s with exit code 1]
[cmd: ['/usr/local/bin/pio', '-f', '-c', 'sublimetext', 'run', '--target', 'upload']]
```

At which point I thought it would be quicker to manually patch as per https://github.com/ieb/ModbusShunt#megatinycore

Upgrading to 2.6.4 fixed the problem I was seeing (first 7 bytes of a modbus RTU frame being dropped) which are more likely to have been caused by a bug in my code than anything in 2.5.11.

**Reply by ieb** [reply]
bug in my code fixed. 2.5.11 works just fine,  but I would still like to find out how to use 2.6.4 or later releases via a recommended pattern.

**Comment by Morcatko** [comment]
If you open the platformio registry page (https://registry.platformio.org/tools/platformio/framework-arduino-megaavr-megatinycore) You can see "Used By" section in the right bar.
It shows "atmelmegaavr" which has source code here https://github.com/platformio/platform-atmelmegaavr. This is probably the palce where the other files come from.

Unfortunatelly I don't know how to force platformio to use last version of megaTinyCode (It shows 2.5.11 for me as well)

**Comment by SpenceKonde** [comment]
And you think I know?!

I wouldn't know platform-io from a hole in the ground! I only know how to
use the arduino ide and how to make a modern avr walk on its side in
assembly :-pme to do?

My brain only has 16 bits of address space anyway (ever wonder why i never
talk about the m2560? Above 64k words of flash, I'm out of my league!). And
here you are asking about some piece of software for desktop pc's with GBs
of ram and TBs of nonvolatile memory - what am I supposed to do?!


____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by mairas** [comment]
LOL, poor Spence and his severely constrained wetware...

I am at the moment struggling with the same issue. PIO comes with an ancient (relatively speaking) megaTinyCore and I really wanted the latest one with its I2C fanciness. I did manage to do a local upgrade. Here goes:

First, clone `megaTinyCore` locally. Then, copy the following snippet into `megaavr/package.json`:

```
 {
  "name": "framework-arduino-megaavr-megatinycore",
  "version": "2.6.5",
  "description": "Arduino Wiring-based Framework for Microchip tinyAVR 0-series and 1-series chips (megaTinyCore)",
  "keywords": [
    "framework",
    "arduino",
    "microchip",
    "tinyAVR"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/SpenceKonde/megaTinyCore"
  }
}
```

Finally, make your project use the updated package by placing the following in your `platformio.ini`:

```
platform_packages =
    framework-arduino-megaavr-megatinycore @ file:///Users/mairas/src/SpenceKonde/megaTinyCore/megaavr
```

Adapt the path as required...

This should be enough to make PIO use the new framework. Additionally, you'll want to get the latest `avrdude` for its `serialupdi` greatness. I tried to build it myself but it turned out to be a hassle. Nightly binaries can be downloaded from here: https://nightly.link/avrdudes/avrdude/workflows/build/main (with your own responsibility! Never run random binaries, etc etc!) Copy the uncompressed files to `/Users/mairas/.platformio/packages/tool-avrdude-megaavr` and PIO should never notice what just happened.

Once you have done all this, you can use `platformio.ini` definitions such as:

```
; Chip in use
board = ATtiny1616
; Clock frequency in [Hz]
board_build.f_cpu = 10000000L
; Oscillator type (internal or external)
board_hardware.oscillator = internal

upload_protocol = serialupdi
```

and it just works!

**Comment by mairas** [comment]
Additionally, I created the following issue: https://github.com/platformio/platform-atmelmegaavr/issues/53

Go give it thumbs up, and if you want to follow the progress, subscribe to it!

EDIT: Here's a respective issue for `avrdude`: https://github.com/platformio/platform-atmelmegaavr/issues/46
